legacy/firmware/messages.c

```diff
@@ -301,32 +308,49 @@ const uint8_t *msg_debug_out_data(void) {
 
 CONFIDENTIAL uint8_t msg_tiny[128];
 _Static_assert(sizeof(msg_tiny) >= sizeof(Cancel), "msg_tiny too tiny");
+_Static_assert(USB_PACKET_SIZE >= MSG_HEADER_SIZE + Cancel_size,
+               "msg_tiny too tiny");
 _Static_assert(sizeof(msg_tiny) >= sizeof(Initialize), "msg_tiny too tiny");
+_Static_assert(USB_PACKET_SIZE >= MSG_HEADER_SIZE + Initialize_size,
+               "msg_tiny too tiny");
 _Static_assert(sizeof(msg_tiny) >= sizeof(PassphraseAck), "msg_tiny too tiny");
+_Static_assert(USB_PACKET_SIZE >= MSG_HEADER_SIZE + PassphraseAck_size,
+               "msg_tiny too tiny");
 _Static_assert(sizeof(msg_tiny) >= sizeof(ButtonAck), "msg_tiny too tiny");
+_Static_assert(USB_PACKET_SIZE >= MSG_HEADER_SIZE + ButtonAck_size,
+               "msg_tiny too tiny");
 _Static_assert(sizeof(msg_tiny) >= sizeof(PinMatrixAck), "msg_tiny too tiny");
+_Static_assert(USB_PACKET_SIZE >= MSG_HEADER_SIZE + PinMatrixAck_size,
+               "msg_tiny too tiny");
 #if DEBUG_LINK
 _Static_assert(sizeof(msg_tiny) >= sizeof(DebugLinkDecision),
                "msg_tiny too tiny");
+_Static_assert(USB_PACKET_SIZE >= MSG_HEADER_SIZE + DebugLinkDecision_size,
+               "msg_tiny too tiny");
 _Static_assert(sizeof(msg_tiny) >= sizeof(DebugLinkGetState),
                "msg_tiny too tiny");
+_Static_assert(USB_PACKET_SIZE >= MSG_HEADER_SIZE + DebugLinkGetState_size,
+               "msg_tiny too tiny");
 #endif
 uint16_t msg_tiny_id = 0xFFFF;
 
 void msg_read_tiny(const uint8_t *buf, int len) {
-  if (len != 64) return;
-  if (buf[0] != '?' || buf[1] != '#' || buf[2] != '#') {
-    return;
+  if (len != USB_PACKET_SIZE || buf[0] != '?' || buf[1] != '#' ||
+      buf[2] != '#') {
+    fsm_sendFailure(FailureType_Failure_UnexpectedMessage,
+                    _("Unknown message"));
+    msg_tiny_id = 0xFFFF;
```

Comment:
Original code did not throw error by design. If unexpected message was pushed to Trezor it was silently skipped. This has proven to be helpful when Trezor was used from two Wallet tabs or similar cases.

We can discuss this on the daily standup.

Reply:
That makes sense, but the error handling seems inconsistent to me. Let's go through the individual checks and let's see which ones are there to actually achieve that goal and which ones are contrary to the goal:
- `len != 64`: I think this should really be `len < MSG_HEADER_SIZE`, but let's leave as is, because it's also a sort of sanity check. Packet should be ignored, I guess.
- `buf[0] != '?' || buf[1] != '#' || buf[2] != '#'`: Continuation packet or malformed packet. Continuation packet should be ignored. Not sure about malformed packet (`buf[0] != '?'`). I suppose we can just leave the current behavior, since it has worked well for us until now.
- `msg_size > 64`: I think it was meant to catch messages which wouldn't fit into `msg_tiny`. Still makes sense to have this as a failsafe in case we forget to add a static assert above. However, it should probably be just before `pb_decode()` and should raise a data error there, because for the unexpected messages we don't care about this check and for the expected ones this shouldn't happen.
- `len - msg_size < 9`: The protobuf message does not fit into one USB packet. Again, this seems like it should be just before `pb_decode()` and should raise a data error there. Same reasoning as the previous point.
- The `else` branch of `if (fields)`: This branch returns `FailureType_Failure_UnexpectedMessage`, which totally contradicts the reasoning about ignoring unexpected messages. Packet should be ignored. In fact there should be a `default` branch in the `switch` statement which will return.

Reply:
For the record, a summary of the two main behavioral issues this solves:
- When a message was unexpected and small enough, it raised a DataError, which contradicts the reasoning about ignoring unexpected messages. Now all unexpected messages are ignored regardless of their size.
- When a message was expected, but too big, it was ignored, which makes debugging difficult. Now it raises a DataError.